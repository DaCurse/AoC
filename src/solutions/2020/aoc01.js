export function partOne(inputs) {
  // We will store numbers in a set until we find our solution
  const prev = new Set();
  const target = 2020;

  for (let i = 0; i < inputs.length; ++i) {
    const input = inputs[i];

    // If a number exists in the set who is the difference between the target
    // and the current number, we found our pair
    if (!prev.has(target - input)) {
      prev.add(input);
    } else {
      return {
        // Returning the pair alongisde the result for propery testing
        fluff: [target - input, input],
        result: (target - input) * input,
      };
    }
  }
}
