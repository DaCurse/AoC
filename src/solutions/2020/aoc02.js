const splitRegex = /-|\s|: /;

/**
 * @param {string[]} inputs Array of policies
 */
export function partOne(inputs) {
  return inputs.reduce((acc, input) => {
    // Parse line using regex
    const [min, max, letter, pw] = input.split(splitRegex);
    // Count occurances
    const count = pw.split(letter).length - 1;

    // Incrementing accumulator if count is within the range
    return count >= Number(min) && count <= Number(max) ? acc + 1 : acc;
  }, 0);
}

/**
 * @param {string[]} inputs Array of policies
 */
export function partTwo(inputs) {
  return inputs.reduce((acc, input) => {
    // Parse line using regex
    const [pos1, pos2, letter, pw] = input.split(splitRegex);

    // Use XOR to determine if only one of the statements is true and incremenet
    // accumulator accordingly
    return (pw[Number(pos1) - 1] === letter) ^ (pw[Number(pos2) - 1] === letter)
      ? acc + 1
      : acc;
  }, 0);
}
