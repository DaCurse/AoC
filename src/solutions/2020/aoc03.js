export function partOne(inputs) {
  const lineLength = inputs[0].length;
  let x = 0;
  let y = 0;
  let trees = 0;
  while (y < inputs.length - 1) {
    x = (x + 3) % lineLength;
    if (inputs[++y][x] === '#') {
      trees++;
    }
  }

  return trees;
}
