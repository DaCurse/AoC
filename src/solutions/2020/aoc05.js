const map = {
  B: '1',
  R: '1',
  F: '0',
  L: '0',
};

function calcIDs(inputs) {
  return inputs.map((input) =>
    parseInt(
      input.replace(/[BR]|[FL]/g, (char) => map[char]),
      2
    )
  );
}

export function partOne(inputs) {
  return Math.max(...calcIDs(inputs));
}

export function partTwo(inputs) {
  const xor = (a, b) => a ^ b;
  const ids = calcIDs(inputs);
  const max = Math.max(...ids);
  const min = Math.min(...ids);

  return (
    ids.reduce(xor) ^
    Array.from({ length: max }, (_, i) => i + 1).reduce(xor) ^
    Array.from({ length: min - 1 }, (_, i) => i + 1).reduce(xor)
  );
}
