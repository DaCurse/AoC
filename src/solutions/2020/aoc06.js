export function partOne(inputs) {
  return inputs.reduce(
    (total, input) => total + new Set(input.replace(/\s/g, '')).size,
    0
  );
}
