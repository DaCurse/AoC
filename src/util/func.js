/**
 * Returns the sum of `n1` and `n2`
 * @param {number} n1
 * @param {number} n2
 */
export function sum(n1, n2) {
  return n1 + n2;
}

/**
 * Returns the product of `n1` and `n2`
 * @param {number} n1
 * @param {number} n2
 */
export function product(n1, n2) {
  return n1 * n2;
}

/**
 * Returns the difference between `n1` and `n2`
 * @param {number} n1
 * @param {number} n2
 */
export function diff(n1, n2) {
  return n1 - n2;
}

/**
 * Returns the quotient from dividing `n1` by `n2`
 * @param {number} n1
 * @param {number} n2
 */
export function quotient(n1, n2) {
  return n1 / n2;
}
