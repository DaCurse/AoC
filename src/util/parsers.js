/**
 * Parses a list of strings to an array with a newline as the delimiter, getting
 * rid of empty results
 * @param {string} raw List of strings
 */
export function parseListOfStrings(raw) {
  return raw.split(/\r?\n/).filter((s) => s);
}

/**
 * Parses a list of groups seperated each by a blank line
 * @param {string} raw List of groups
 */
export function parseListOfGroups(raw) {
  return raw.split(/(?:\r?\n){2}/).filter((s) => s);
}

/**
 * Parses list of strings to an array of integers
 * @param {string} raw List of strings
 */
export function parseListOfInts(raw) {
  return parseListOfStrings(raw).map((n) => parseInt(n, 10));
}
