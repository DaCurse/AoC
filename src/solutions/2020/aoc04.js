import { parseListOfGroups } from '../../util/parsers';

/**
 * This abomination parses a list of passports (huge string) seperated by blank lines into objects with keys/values
 * being each field and it's respective value
 * @param {string} raw Raw input (List of passports)
 */
export function parsePassports(raw) {
  return parseListOfGroups(raw).map((passport) => {
    const parts = passport.split(/\s/);
    return parts.reduce((obj, part) => {
      const [key, value] = part.split(':');
      if (key === 'cid' || key.length === 0) return obj;
      return { ...obj, [key]: value };
    }, {});
  });
}

/**
 * Checks if a passport has all required fields
 * @param {string[]} fields Array of fields
 */
function hasAllFields(fields) {
  const requiredFields = ['byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid'];
  return requiredFields.every((field) => fields.includes(field));
}

export function partOne(inputs) {
  // Return the count of all passports with all required fields
  return inputs.filter((passport) => hasAllFields(Object.keys(passport)))
    .length;
}

/**
 * Utility function to convert `value` to String and check if it's within a
 * certain range
 * @param {string} value
 * @param {number} min
 * @param {number} max
 */
function between(value, min, max) {
  return parseInt(value) >= min && parseInt(value) <= max;
}

// Key/value pairs where the key is the field and the value is a function which check if it's valid
const predicates = {
  byr(value) {
    return between(value, 1920, 2002);
  },
  iyr(value) {
    return between(value, 2010, 2020);
  },
  eyr(value) {
    return between(value, 2020, 2030);
  },
  hgt(value) {
    return value.endsWith('cm')
      ? between(value, 150, 193)
      : value.endsWith('in')
      ? between(value, 59, 76)
      : false;
  },
  hcl(value) {
    return /^#[0-9a-f]{6}$/.test(value);
  },
  ecl(value) {
    return ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth'].includes(value);
  },
  pid(value) {
    return /^[0-9]{9}$/.test(value);
  },
};

export function partTwo(inputs) {
  // Return the count of all passports that have all required fields and field values pass all the predicates
  return inputs.filter(
    (passport) =>
      hasAllFields(Object.keys(passport)) &&
      Object.keys(passport).every((field) => predicates[field](passport[field]))
  ).length;
}
