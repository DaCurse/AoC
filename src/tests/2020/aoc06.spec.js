import { partOne, partTwo } from '../../solutions/2020/aoc06';
import getInputs from '../../util/inputs';
import { parseListOfGroups } from '../../util/parsers';

describe('Day 6 solution', () => {
  test('Part one: Find the sum of questions for each group', async () => {
    const inputs = await getInputs(6, parseListOfGroups);
    const result = partOne(inputs);
    console.log('Day 6 Part 1 solution:', result);

    expect(result).toMatchSnapshot();
  });

  test('Part one: Find the sum of unanimous questions for each group', async () => {
    const inputs = await getInputs(6, parseListOfGroups);
    const result = partTwo(inputs);
    console.log('Day 6 Part 2 solution:', result);

    expect(result).toMatchSnapshot();
  });
});
