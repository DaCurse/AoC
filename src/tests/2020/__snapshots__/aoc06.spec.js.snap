// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day 6 solution Part one: Find the sum of questions for each group 1`] = `6633`;
